### PR TITLE
Add desktop menu dismiss rate P3A metric

### PR DESCRIPTION
**Pull request by DJAndries**
Resolves brave/brave-browser#26913

## Submitter Checklist:

- [ ] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

0. Start with fresh profile, ensure metric has value of 0.
1. Open the menu and open a new tab 4 times. Ensure the metric has a value of 1.
2. Open the menu and dismiss it 2 times. Ensure the metric has a value of 2.
3. Open the menu and dismiss it 2 more times. Metric should equal 3.
4. Open the menu and dismiss it 2 more times. Metric should equal 4.
5. Fast forward system time by 7 days and restart browser, ensure metric resets to 0.

### CODE REVIEW

**Comment by DJAndries** [comment]
Ignore first two commits, they are part of #15876

File: components/misc_metrics/menu_metrics.h [hunk 1]
```diff
@@ -33,10 +33,21 @@ class MenuMetrics {
 
   void RecordMenuGroupAction(MenuGroup group);
 
+  void RecordMenuShown();
```
Comment (iefremov) [inline components/misc_metrics/menu_metrics.h:36]: pls add comments

File: components/misc_metrics/menu_metrics.h [hunk 2]
```diff
@@ -33,10 +33,21 @@ class MenuMetrics {
...
+
+  void Update();
+
   base::flat_map<MenuGroup, double> menu_group_access_counts_;
 
   raw_ptr<PrefService> local_state_ = nullptr;
+  WeeklyStorage menu_shown_storage_;
+  WeeklyStorage menu_dismiss_storage_;
+
+  base::RepeatingTimer update_timer_;
```
Comment (iefremov) [inline components/misc_metrics/menu_metrics.h:50]: you need `WallClockTimer` or a smaller interval

File: components/misc_metrics/menu_metrics.cc [hunk 3]
```diff
@@ -93,4 +104,44 @@ void MenuMetrics::RecordMenuGroupAction(MenuGroup group) {
...
+}
+
+void MenuMetrics::RecordMenuDismiss() {
+  VLOG(2) << "MenuMetrics: menu dismiss";
+  menu_dismiss_storage_.AddDelta(1);
+  RecordMenuDismissRate();
+}
+
+void MenuMetrics::RecordMenuDismissRate() {
+  double shown_sum = menu_shown_storage_.GetWeeklySum();
```
Comment (iefremov) [inline components/misc_metrics/menu_metrics.cc:120]: throwing in few `const` never hurts

File: components/misc_metrics/menu_metrics.h (hunk 2 above)
Comment (DJAndries) [reply components/misc_metrics/menu_metrics.h:50]: can you pls elaborate on why a wall clock timer would be needed here?

Comment (iefremov) [reply components/misc_metrics/menu_metrics.h:70]: it seems to me that it is always needed when the interval is relatively large, say more than an hour or two? apparently most of devices can't avoid sleeping during a day

**Review by iefremov — APPROVED** [review]

File: components/misc_metrics/menu_metrics.h (hunk 2 above)
Comment (DJAndries) [reply components/misc_metrics/menu_metrics.h:70]: good point. there are quite a few places where we use repeating timers with daily intervals (i.e. wallet). we should revisit those cases at some point

**Pull request merged**